File: LeetCode/Linked list/p92/c++/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode *p = head,*p1 = new ListNode();
        stack<int> list;
        int n = 0;
        while(++n<=right){
            if(n==left){
                p1 = p;
            }
            if(n>=left){
                list.push(p->val);
            }
            p = p->next;
        }
        n -= left;
        while(n--){
            p1->val = list.top();
            list.pop();
            p1 = p1->next;
        }
        return head;
    }
};

File: LeetCode/Linked list/p92/java/solution.java
class Solution {
    ListNode tail = null;
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(left == 1) {
            return reverse(head,right);
        }
        head.next = reverseBetween(head.next,left-1,right-1);
        return head;
    }
    public ListNode reverse(ListNode head,int n) {
        if(n==1) {
            tail = head.next;
            return head;
        }
        ListNode last = reverse(head.next,n-1);
        head.next.next = head;
        head.next = tail;
        return last;
    }
}